swarm/storage/memstore_lrucache.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2018 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+// memory storage layer for the package blockhash
+
+package storage
+
+import (
+	"bytes"
+	"fmt"
+	"sync"
+
+	lru "github.com/hashicorp/golang-lru"
+)
+
+type MemStore struct {
+	cache    *lru.Cache
+	requests *lru.Cache
+	mu       sync.Mutex
+	disabled bool
+}
+
+func NewMemStore(_ *LDBStore, cacheCapacity uint, requestsCapacity uint) (m *MemStore) {
+	if cacheCapacity == 0 {
+		return &MemStore{
+			disabled: true,
+		}
+	}
+
+	onEvicted := func(key interface{}, value interface{}) {
+		v := value.(*Chunk)
+		<-v.dbStoredC
+	}
+	c, err := lru.NewWithEvict(int(cacheCapacity), onEvicted)
+	if err != nil {
+		panic(err)
+	}
+
+	r, err := lru.New(int(requestsCapacity))
+	if err != nil {
+		panic(err)
+	}
+
+	return &MemStore{
+		cache:    c,
+		requests: r,
+	}
+}
+
+func (m *MemStore) Get(key Key) (*Chunk, error) {
+	if m.disabled {
+		return nil, ErrChunkNotFound
+	}
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	r, ok := m.requests.Get(string(key))
+	// it is a request
+	if ok {
+		return r.(*Chunk), nil
+	}
+
+	// it is not a request
+	c, ok := m.cache.Get(string(key))
+	if !ok {
+		return nil, ErrChunkNotFound
+	}
+	chunk := c.(*Chunk)
+	if !bytes.Equal(chunk.Key, key) {
+		panic(fmt.Errorf("MemStore.Get: chunk key %s != req key %s", chunk.Key.Hex(), key.Hex()))
+	}
```

Comment:
I'm not sure why we need this check, but if we need it, don't we need it in the case when the chunk is retrieved from the request cache? Becuase in that case this check does not run, we return earlier.

Reply:
I was wondering the same. I am not sure in what scenario this is even possible, I kept it as it was already here.

Reply:
I would just remove this, it is a panic anyway which should never happen. I guess this was added when somebody debugged an issue. But I'm ok with keeping it too, not a big deal

Reply:
Removed.